### PR TITLE
Add functionality to replace existing templates

### PR DESCRIPTION
**Pull request by joeyberkovitz**
Adds functionality to delete any existing VMs with the same name as the `template_name` variable
Should resolve https://github.com/hashicorp/packer-plugin-proxmox/issues/76

To support this change, the go.mod was updated because the build was failing due to an older google cloud dependency that was now broken. As part of the upgrade, the upstream Proxmox API EFI definition changed from a string to an object. I just updated that config to match the definition in the upstream README.

### CODE REVIEW

**Comment by goffinf** [comment]
This sounds like it adds useful functionality. I appreciate that you have set the default is 'false', however, can I ask what the behaviour is when replacing a template where there are existing linked-clones. From the docs:-

> You cannot delete an original template while linked clones exist.

Thanks

Fraser

**Comment by joeyberkovitz** [comment]
It was my understanding that the source template can be removed without impacting the linked clone. Based on https://forum.proxmox.com/threads/i-deleted-original-template-while-linked-clones-exist.94879/ - it sounds like a linked clone is essentially a hard link, where if the source is deleted, the file still exists until all dependent VMs are also removed.

If this is not the case, I can look into what Proxmox actually does and potentially look into providing some sort of functionality to convert the linked clones into full clones, or maybe just auto-rename the old source VMs. For my use case and that mentioned in the linked issue, it would be ideal to remove the template so that it can be automatically replaced by a CI/CD pipeline without having to manually go in and delete old components.

**Comment by goffinf** [comment]
Hey Joey, thanks for your reply. After a bunch of googling, including the link you provided to the question raised in the forum, I can't find a definitive answer as to the impact. Obviously I would assume that Proxmox's own documentation is correct, however that is not to say what the side effects of deleting an 'in-use' template are. The doc link I was quoting from is here ...

https://proxmox.local.goffinf.co.uk/pve-docs/chapter-qm.html#qm_copy_and_clone

This bears some further investigation IMHO. I can help out if you want (create a template and a linked clone, (attempt to) delete the template (with a VM is still referencing it - and if that fails, stop the VM and attempt the delete - then try to restart ......) ...

I am completely OK if you prefer to conduct any investigation yourself ?

The results of this should probably inform what approach should be taken if there are any destructive side-effects.

I also build all my templates and provision all associated infrastructure via CI pipelines so I agree this need to support automation where it may not be possible for the builder itself to request confirmation (that might have to be left to the pipeline implementation) even if that means there needs to be a prominent warning in the docs about potential side effects (assuming that anyone reads docs these days !).

**Comment by joeyberkovitz** [comment]
I can do some more extensive testing later - either later today or over the next few days, but if you have time to verify as well, that would be helpful. 

I did briefly test this functionality as built in the PR, so I know it lets you delete the template, and I did have one linked clone running, but I didn't do any extensive testing to verify that the linked clone was not impacted in any way.

**Comment by goffinf** [comment]
Yes, I did a similar test, created a linked clone (with lvm-thinpool storage) from a template, then successfully deleted the template on which it was based, while it was still running. 

Given the code link you provided above I had expected this to fail ... what am I missing ?

**Comment by nywilken** [comment]
@carlpett do you have bandwidth to provide guidance on these changes?

**Comment by nywilken** [comment]
> To support this change, the go.mod was updated because the build was failing due to an older google cloud dependency that was now broken. 

@joeyberkovitz the module need to be updated to support this change? Or is it currently broken? 

Looking at the go.mod it is not immediately apparent which module you are referring to. Can you please provide insight to this change as well. Thanks!

**Comment by joeyberkovitz** [comment]
@nywilken  - when I pulled the code a few days ago, it wasn't in a state where it would build. I was getting some sort of build error related to one of the following modules:
```
cloud.google.com/go v0.94.0 // indirect
cloud.google.com/go/storage v1.16.1 // indirect
```

I didn't do any analysis as to what was including those modules and instead ran a go mod update, which pulled in some new API changes from the upstream Proxmox repo.

**Comment by nywilken** [comment]
> @nywilken - when I pulled the code a few days ago, it wasn't in a state where it would build. I was getting some sort of build error related to one of the following modules:
> 
> ```
> cloud.google.com/go v0.94.0 // indirect
> cloud.google.com/go/storage v1.16.1 // indirect
> ```
> 
> I didn't do any analysis as to what was including those modules and instead ran a go mod update, which pulled in some new API changes from the upstream Proxmox repo.
> 
> If I remember correctly, this was the error: [googleapis/google-cloud-go#5304](https://github.com/googleapis/google-cloud-go/issues/5304)

Thanks for the quick reply. Could you provide the version of Go being used to build. I would like to try and reproduce locally. I'm not seeing this issue when building with Go 1.17.8.

**Comment by joeyberkovitz** [comment]
I'm on Go version 1.18.3.

If the google cloud problem is a non-issue, I'd be happy to revert that change along with the associated Proxmox updates, as it's not really related to this PR.

**Comment by xoxys** [comment]
Why do we need a new plugin option `replace_existing` for this functionality? Shouldn't plugins the default `-force` flag from the packer build command?

**Comment by joeyberkovitz** [comment]
Sure - we can use `-force`, the only difference as compared to other hypervisors such as VMware is that VM names aren't guaranteed to be unique, so force here would potentially delete multiple VMs. If that sounds reasonable, it's not a big deal to switch it over.

**Comment by xoxys** [comment]
Can we prefere the ID if set and only fallback to the name? It would IMO also be reasonable to use `force` only if the ID is set to avoid the issue with multiple names.

**Comment by joeyberkovitz** [comment]
sure, that could be done
personally, I don't think I would ever use it that way - carefully managing VM IDs seems a bit clunky as compared to just ensuring that a given VM name exists, but I wouldn't have any issue building the logic for that.

**Comment by joeyberkovitz** [comment]
latest commit uses `-force` instead of adding `replace_existing`. Also supports replacing a single VM/template by ID if ID is specified, otherwise it'll replace by name

also added unit tests which verify that delete is never called if force is disabled, and that delete is called appropriately if force is set

**Comment by modem7** [comment]
Would the EFI changes conflict with https://github.com/hashicorp/packer-plugin-proxmox/pull/90?

**Comment by joeyberkovitz** [comment]
Yes, it probably makes sense to merge #90 first then I can rebase to that

**Comment by sebastian-de** [comment]
Hey @joeyberkovitz
now that the EFI-Chnges are merged, I have rebased your changes on the current main branch: https://github.com/sebastian-de/packer-plugin-proxmox/tree/replace

Feel free to use this branch if you'd like to update your PR.

**Comment by joeyberkovitz** [comment]
I switched over to your rebased branch. Code looks fine, but for some reason I can't build on my machine - pretty sure this is the same issue I ran into initially which led to upgrading dependencies:

Output from `go build .` after running `go get -u all`

```# cloud.google.com/go/storage
...\go\pkg\mod\cloud.google.com\go\storage@v1.16.1\storage.go:1416:54: o.GetCustomerEncryption().GetKeySha256 undefined (type *"google.golang.org/genproto/googleapis/storage/v2".CustomerEncryption has no field or method GetKeySha256)```

**Comment by sebastian-de** [comment]
Hmm, I can build this branch here without problems. Which go version do you use? I currently use `go1.19.4`.

**Comment by joeyberkovitz** [comment]
I tested on `1.19.5` and also reset my `go/pkg` directory followed by `go build .` and get the same error. not sure if this is a Windows only issue, if so, I guess it's fine.

**Comment by sebastian-de** [comment]
Hey @joeyberkovitz I tested your changes today and have some suggestions.
In general your approach works fine and existing VMs are removed when `packer build -force` us used. That's definitely a much needed feature for this builder - thanks for your work on this!

However in my opinion Packer should be more strict on what VMs it'll choose to remove. Citing the [Packer docs](https://github.com/hashicorp/packer/blob/main/website/content/docs/commands/build.mdx) what `-force` should do:
> Forces a builder to run when artifacts from a previous build prevent a build from running.

We have no way to check if the specified VM was created by Packer, but we can make a few assumptions. One is that the Proxmox builders always produce templates. So I think we should check if the given VM is a template before removing it. Since template VMs are never in a `running` state, we don't have to stop VMs either. In my opinion Packer should never attempt to stop a running VM.

Another thing that should be checked before removing a VM is that there exists only one VM with the given `vm_name`. As @xoxys already said - VM names aren't unique in a Proxmox cluster, so we should be extra cautious when names are used to delete VMs.

I think with these changes `-force` should still be compatible with your use case.

I created a commit to show what I mean here: https://github.com/sebastian-de/packer-plugin-proxmox/commit/4cbb6bdd5987d176d4ddb022e7edcceefefc3031 (tests not adjusted yet).
Please let me know what you think!

**Comment by xoxys** [comment]
To be honest, this change is rather small. The default behavior will not change, and the use of an additional `-force` flag already adds an extra layer of protection in my opinion. Even if the behavior can be improved, I really wonder a bit why such a change has not been merged for ~6 months.

**Comment by joeyberkovitz** [comment]
@sebastian-de: I'm fine with adjusting the logic to only remove templates if that is the most desirable design

With the current design, if an ID is specified, it's pretty safe as it will only delete at most one VM with the specified ID, creating a new one to replace it.

The idea with deleting any VM matching a given name was that at some later point you're probably going to want to clone that template into a VM, at which point it could be important to have that name be unique.

**Review by lbajolet-hashicorp — COMMENTED** [review]
Hi @joeyberkovitz,

Sorry for letting this PR slide under the radar for so long, thank you for updating it during all this time.

I've taken a look at the code here, and on @sebastian-de's branch as they linked it in a discussion above.

I think they have a point regarding whether or not we should stop running VMs that correspond to the template we are re-building, if the template is the thing we are replacing, we can probably keep the existing VMs alive so we can decide later if we want to stop them.

On that note I must say that I'm not necessarily a Proxmox expert and I don't have any infrastructure to test those changes on, so ultimately I'll leave it to you to decide if that's what the plugin should do or not.

Apologies again for the delay in reviewing this PR, I'll follow this PR closely, and will try to get it merged as soon as we're all satisfied with the code.

**Comment by sebastian-de** [comment]
> I tested on `1.19.5` and also reset my `go/pkg` directory followed by `go build .` and get the same error. not sure if this is a Windows only issue, if so, I guess it's fine.

Regarding your build problems: Have you tried updating only the dependency that causes the error?
`go get cloud.google.com/go/storage && go mod tidy`

**Comment by joeyberkovitz** [comment]
yes, that works, although a bunch of indirect dependencies end up being updated as a result

File: builder/proxmox/common/step_start_vm.go
Comment (joeyberkovitz) [reply builder/proxmox/common/step_start_vm.go:95]: yes, although it's a pretty sub-optimal check. the proxmox client doesn't have typed errors, so we would have to check if the error string matches either `vm '%d' not found` or `vm '%s' not found` with the VM ID or name filled in.

I have no problem adding this with some logic to just note when nothing is found, but error and fail if there's a different error, although this could break if the client is updated and the error string ever changes

File: builder/proxmox/common/step_start_vm.go
Comment (joeyberkovitz) [reply builder/proxmox/common/step_start_vm.go:107]: sure. that could be an option

**Comment by joeyberkovitz** [comment]
@sebastian-de - I ultimately have nothing against only deleting VMs that are templates if that is the preferable route, just let me know if that's the decided on option and I can switch over to your commit and update the tests.

**Comment by sebastian-de** [comment]
Hey @joeyberkovitz, please excuse me if my comment sounded like I had already made a decision on the matter. I am also still open discuss whether we should delete VMs that are not templates.
Since we plan to use Packer on our production clusters, I'm just leaning towards a more conservative approach.
But of course I would be happy if you like to include my commit - the updated version with adjusted tests is here: https://github.com/sebastian-de/packer-plugin-proxmox/commit/1a1faab5bc1849de1a5ef43e97a786fa43087579

**Comment by joeyberkovitz** [comment]
I see the reasoning for preferring to not delete multiple VMs or stop any running VMs. I think that in a more realistic scenario where Packer is running more continually, updating the same image, there would only ever be exactly one template with a given name, and as a result it would never be running.

When a person onboards onto Packer though, they would have to manually cleanup the Proxmox cluster if there are any dups or running VMs with a name matching the template. This scenario is what I came across frequently during testing, although once this feature is merged and a cluster is manually cleaned up, if Packer is the only one ever creating templates, then it should never have to really have to delete running or non-template VMs.

Overall, I think the extra deletions could be nice for that onboarding, but given that it's usually going to be a one-time task, we could just not support that use case and lean toward the conservative option.

**Comment by sebastian-de** [comment]
> When a person onboards onto Packer though, they would have to manually cleanup the Proxmox cluster if there are any dups or running VMs with a name matching the template.

I get why you want to avoid that, but I remain cautious because in that scenario it's possible that Packer may delete VMs that the user did not intend to. IMHO that would be a far worse user experience.
If it's implemented that way we should at least add a big warning to the docs that using `-force` together with `vm_name` will delete _all_ VMs matching that name.

**Comment by joeyberkovitz** [comment]
Understood and agreed that it makes sense to be more cautious at the expense of potentially a bit more work during initial onboarding.

As such, I've pulled in the changes from @sebastian-de. Both comments from @lbajolet-hashicorp should be resolved - the changes include a check to see what error was returned from proxmox (although it can fail if the proxmox API client error strings change in the future). The timeout is no longer relevant since we'll never stop any running VMs

**Comment by xoxys** [comment]
During the development of my packer templates, it happened quite often that the VM creation/initialization has not finished. As a result, the final template was never created and the VM still exist. To restart the build process, it's not required to manually clean up all remaining components __every time__.

That's how I came across this PR, and I think using an extra flag like `-force` with a proper documentation is safe. Now it ha moved in a direction where this use case is not covered anymore and people would still have to clean up stuff manually because it's considered even more safe. What would be the suggested implementation to do it automatically? Adding another flag `-force-all`?

**Comment by joeyberkovitz** [comment]
This becomes a question of how safe the inherently unsafe force operation should be, given that it needs to be used in production scenarios. Below are some options:

- unsafe: stop and delete all VMs matching the name or ID if specified
- semi-safe: relax the template requirement if ID is specified to allow for stopping a single VM
- semi-safe: don't ever check if a VM is a template and always stop and delete at most 1 VM
- safe: as is right now, only delete at most 1 template

Another option is to optionally delete a VM on failure like the VMware builder does, although if you needed to debug, you would likely disable that feature anyway. 

Considering the development use case, I would prefer one of the semi-safe or unsafe options. If needed flags can also be added to allow for deleting non templates and stopping running VMs. 

Once a decision is made regarding how the plugin should be designed, I can adjust accordingly.

**Comment by lbajolet-hashicorp** [comment]
I think @joeyberkovitz hits the nail with the "delete an image in case of failure".

Generally plugins are responsible for cleaning-up after them, in this case that would mean deleting the VM in case of error.

To your point regarding debugging, there is a flag on Packer that you can use to specify what to do in case of error `--on-error`, which defaults to clean-up, but if you wanted to wait and ssh into the VM, you could set it to `ask`, where it will wait until you choose what to do with the VM.
Unless I'm mistaken, `abort` would leave the VM alive as well, it would abort immediately the build without cleaning-up anything.

The `--force` flag is part of Packer, and its semantics already depend on what the plugins want to do with it, as such I don't think we should have another `--force-all` flag, I don't think the project will benefit from that. On the contrary even, it will add another layer of complexity and confusion, so I think we should steer clear from this.

Note that this doesn't need to be part of this PR, it can be opened as a subsequent one later.
Judging from the discussion, automatically removing the VM at the end of the process (or only in case of failure? That depends how we implement the cleanup) combined with the `--on-error` flag would work well , what does everyone think?

**Comment by sebastian-de** [comment]
I never used the `--on-error` flag, but without it Packer removes failed builds pretty reliably for me, even if it's killed via keyboard-interrupt. So `Cleanup()` seems to work fine in the majority of cases.
https://github.com/hashicorp/packer-plugin-proxmox/blob/b91ce000718075bf5385c5598a963c0fe4579f6b/builder/proxmox/common/step_start_vm.go#L253

The only occasion I had a leftover VM was when I killed our GitLab-runner which was taking down the Packer-container with it, so Packer had no chance to clean up. After that I had to stop and delete the VM manually.

So I get why @xoxys wants a more aggressive `-force` for situations like this.

With all the different variables, I guess we won't find an easy solution that fits all usecases, so I'm still in favor of the current, least aggressive approach.

**Comment by xoxys** [comment]
All valid points, and I think `--on-error=ask` would be a feasible way to debug and cleanup. Let's continue with the current least aggressive approach :+1:

**Comment by lbajolet-hashicorp** [comment]
With what @sebastian-de shared it seems the plugin does cleanup the VM, I'm wondering why you have to manually cleanup the VMs then @xoxys since it seems the plugin should stop and remove the one being built. Do you forcibly kill Packer or do you let it perform the cleanup when the build fails?

If you forcibly kill it then it makes sense, otherwise, there's probably a bug begging to be addressed in this logic.

**Comment by xoxys** [comment]
To be honest I cant rember the details and would assume I have done something wrong. Sorry for the confusion.

**Review by lbajolet-hashicorp — APPROVED** [review]
@joeyberkovitz, thanks for the multiple rerolls.

Regarding the error message check I do agree this is not super robust as if the message changes on the API, we'll start failing when attempting to delete a VM that doesn't exist, but I think this will be good enough until the API offers a better option.

Unless someone has an objection, I will merge this tomorrow, in its current state the PR looks good to me.

**Pull request merged**
